File: typescript/src/tool/draw.ts
/**
 * The renderer client tools the SDK ships: ``cosmo_sdk_draw_box`` and
 * ``cosmo_sdk_draw_point``.
 *
 * A server-side locator (``{ kind: 'detect_objects' }`` /
 * ``{ kind: 'point_at_object' }``) hands the model candidate boxes or points;
 * the model picks the one that matches what it is looking at and passes it
 * here for the app to draw over the user's live view::
 *
 *     const agent = client.agent({
 *       instructions: 'help the user find things on screen',
 *       tools: [
 *         { kind: 'detect_objects' },
 *         drawBox((request) => {
 *           if (!preview.visible) return notShown('the preview is not on screen');
 *           preview.showBox(request);
 *           return shown;
 *         }),
 *       ],
 *     });
 *
 * The SDK owns the name, description, schema, decode and reply shape; the
 * caller supplies one function of request → outcome, closing over whatever
 * app state it draws with. Coordinates, wording, and the reply shape are
 * cross-SDK contract, pinned by ``sdk-client-tool-vectors.json``.
 */

import type { ClientToolSpec } from '../core/agent';

import { markSdkClientTool } from './sdk_tool';

/** Where to draw, normalized to the frame the model was shown: ``[0,1]``,
 *  top-left origin (y increases downward). */
export type NormalizedBox = {
  x: number;
  y: number;
  width: number;
  height: number;
};

/** Where to mark, in the same normalized space. */
export type NormalizedPoint = {
  x: number;
  y: number;
};

/** One model request to draw a box over the user's live view. */
export type DrawBoxRequest = {
  box: NormalizedBox;
  label?: string;
};

/** One model request to mark a single spot over the user's live view. */
export type DrawPointRequest = {
  point: NormalizedPoint;
  label?: string;
};

/** What a renderer reports back to the model.
 *
 *  Drawing can fail for reasons the model has to hear about — the camera is
 *  off, the preview isn't on screen. Answering "shown" regardless would leave
 *  it talking about something the user cannot see, so a refusal carries a
 *  reason the agent can say out loud. */
export type DrawOutcome = { shown: true } | { shown: false; reason: string };

/** The refusal arm every renderer shares — the camera draws and the screen
 *  highlights alike, since "nothing was drawn, and here is why" carries no
 *  detail either of them needs to qualify. Named so {@link notShown} can serve
 *  both without either outcome type importing the other. */
export type NotShown = { shown: false; reason: string };

/** The annotation is on screen. */
export const shown: DrawOutcome = { shown: true };

/** Nothing was drawn. ``reason`` is model-facing: write it as an instruction
 *  the agent can speak ("the camera is off — ask the user to turn it on"),
 *  not as an error code. */
export function notShown(reason: string): NotShown {
  return { shown: false, reason };
}

/** Wire name shipped in tool-call events; a rename is a wire break. */
export const DRAW_BOX_TOOL_NAME = 'cosmo_sdk_draw_box';
export const DRAW_POINT_TOOL_NAME = 'cosmo_sdk_draw_point';

const DRAW_BOX_DESCRIPTION =
  "Draw a box over the user's live view (their camera or screen preview) around " +
  'something cosmo_detect_objects located — pass a box it returned, normalized to ' +
  'the frame you were shown ([0,1], top-left origin), and an optional short label. ' +
  'Call this after cosmo_detect_objects rather than guessing a box yourself. ' +
  'Visual only — it measures nothing and changes nothing.';

const DRAW_POINT_DESCRIPTION =
  "Mark a single spot on the user's live view (their camera or screen preview) — one " +
  'leaf, one screw, one control — using a point cosmo_point_at_object returned, ' +
  'normalized to the frame you were shown ([0,1], top-left origin), with an optional ' +
  'short label. Call this after cosmo_point_at_object rather than guessing a position ' +
  'yourself. Visual only — it measures nothing and changes nothing.';

const DRAW_BOX_PARAMETERS: Record<string, unknown> = {
  type: 'object',
  properties: {
    box: {
      type: 'object',
      description:
        'Where to draw, normalized to the frame you were shown: [0,1], top-left origin.',
      properties: {
        x: { type: 'number', minimum: 0, maximum: 1 },
        y: { type: 'number', minimum: 0, maximum: 1 },
        width: { type: 'number', minimum: 0, maximum: 1 },
        height: { type: 'number', minimum: 0, maximum: 1 },
      },
      required: ['x', 'y', 'width', 'height'],
    },
    label: {
      type: 'string',
      maxLength: 40,
      description: "Short caption shown on the box, e.g. 'blush here'.",
    },
  },
  required: ['box'],
};

const DRAW_POINT_PARAMETERS: Record<string, unknown> = {
  type: 'object',
  properties: {
    point: {
      type: 'object',
      description:
        'Where to point, normalized to the frame you were shown: [0,1], top-left origin.',
      properties: {
        x: { type: 'number', minimum: 0, maximum: 1 },
        y: { type: 'number', minimum: 0, maximum: 1 },
      },
      required: ['x', 'y'],
    },
    label: {
      type: 'string',
      maxLength: 40,
      description: "Short caption shown beside the marker, e.g. 'this screw'.",
    },
  },
  required: ['point'],
};

const clamp01 = (value: number): number => Math.min(1, Math.max(0, value));

function coordinate(value: unknown): number | null {
  return typeof value === 'number' && Number.isFinite(value)
    ? clamp01(value)
    : null;
}

function label(args: Record<string, unknown>): string | undefined {
  return typeof args.label === 'string' ? args.label : undefined;
}

/** Decode a ``cosmo_sdk_draw_box`` invocation. ``null`` when the box is
 *  absent or malformed — a boundary check on model output, not an invariant.
 *  Coordinates are clamped, so a model that overshoots the frame edge still
 *  yields a drawable box. */
export function parseDrawBoxRequest(
  args: Record<string, unknown>,
): DrawBoxRequest | null {
  const raw = args.box;
  if (typeof raw !== 'object' || raw === null) return null;
  const candidate = raw as Record<string, unknown>;
  const x = coordinate(candidate.x);
  const y = coordinate(candidate.y);
  const width = coordinate(candidate.width);
  const height = coordinate(candidate.height);
  if (x === null || y === null || width === null || height === null) return null;
  return { box: { x, y, width, height }, label: label(args) };
}

/** Decode a ``cosmo_sdk_draw_point`` invocation. Same contract as
 *  {@link parseDrawBoxRequest}. */
export function parseDrawPointRequest(
  args: Record<string, unknown>,
): DrawPointRequest | null {
  const raw = args.point;
  if (typeof raw !== 'object' || raw === null) return null;
  const candidate = raw as Record<string, unknown>;
  const x = coordinate(candidate.x);
  const y = coordinate(candidate.y);
  if (x === null || y === null) return null;
  return { point: { x, y }, label: label(args) };
}

/** The box renderer, ready to add to an agent's ``tools`` alongside the
 *  locator that feeds it. Your handler owns only the drawing — and the honest
 *  answer about whether it happened. Malformed arguments surface to the model
 *  as the call's error without reaching your code. */
export function drawBox(
  onDraw: (request: DrawBoxRequest) => DrawOutcome | Promise<DrawOutcome>,
): ClientToolSpec {
  return markSdkClientTool({
    kind: 'client',
    name: DRAW_BOX_TOOL_NAME,
    description: DRAW_BOX_DESCRIPTION,
    parameters: DRAW_BOX_PARAMETERS,
    handler: async (args) => {
      const request = parseDrawBoxRequest(args);
      if (request === null) {
        throw new Error(
          `${DRAW_BOX_TOOL_NAME}: pass box {x,y,width,height} normalized to [0,1]`,
        );
      }
      return { ...(await onDraw(request)) };
    },
  });
}

/** The point renderer. Same contract as {@link drawBox}, with a
 *  {@link DrawPointRequest}. */
export function drawPoint(
  onDraw: (request: DrawPointRequest) => DrawOutcome | Promise<DrawOutcome>,
): ClientToolSpec {
  return markSdkClientTool({
    kind: 'client',
    name: DRAW_POINT_TOOL_NAME,
    description: DRAW_POINT_DESCRIPTION,
    parameters: DRAW_POINT_PARAMETERS,
    handler: async (args) => {
      const request = parseDrawPointRequest(args);
      if (request === null) {
        throw new Error(
          `${DRAW_POINT_TOOL_NAME}: pass point {x,y} normalized to [0,1]`,
        );
      }
      return { ...(await onDraw(request)) };
    },
  });
}

File: python/src/cosmo_ai/tools/_draw.py
"""The renderer half of the locate-then-draw pair: ``cosmo_sdk_draw_box`` and
``cosmo_sdk_draw_point``.

A server-side locator (:class:`~cosmo_ai.DetectObjectsTool` /
:class:`~cosmo_ai.PointAtObjectTool`) returns candidate boxes or points to the
model; the model picks the one matching what it is looking at and passes it to
the renderer, which draws it over the user's live camera or screen preview::

    from cosmo_ai import DetectObjectsTool
    from cosmo_ai.tools import DrawBoxRequest, DrawOutcome, draw_box

    def on_draw(request: DrawBoxRequest) -> DrawOutcome:
        if not camera.streaming:
            return DrawOutcome(
                shown=False,
                reason="the camera is off — ask the user to turn it on",
            )
        overlay.show(request.box, label=request.label)
        return DrawOutcome(shown=True)

    agent = client.agent(tools=[DetectObjectsTool(), draw_box(on_draw)])

The renderers measure nothing — they carry the model's choice to your UI. The
SDK owns the name, description, schema, decode and reply shape; you own the
drawing, and the honest answer about whether it happened.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Union

from cosmo_ai._internal.protocol import ClientTool
from cosmo_ai.tools._sdk_tools import _SdkClientTool

DRAW_BOX_TOOL_NAME = "cosmo_sdk_draw_box"
"""Wire name shipped in ``tool-invocation`` events; a rename is a wire break."""

DRAW_POINT_TOOL_NAME = "cosmo_sdk_draw_point"
"""Wire name shipped in ``tool-invocation`` events; a rename is a wire break."""

_DRAW_BOX_DESCRIPTION = (
    "Draw a box over the user's live view (their camera or screen preview) "
    "around something cosmo_detect_objects located — pass a box it returned, "
    "normalized to the frame you were shown ([0,1], top-left origin), and an "
    "optional short label. Call this after cosmo_detect_objects rather than "
    "guessing a box yourself. Visual only — it measures nothing and changes "
    "nothing."
)

_DRAW_POINT_DESCRIPTION = (
    "Mark a single spot on the user's live view (their camera or screen "
    "preview) — one leaf, one screw, one control — using a point "
    "cosmo_point_at_object returned, normalized to the frame you were shown "
    "([0,1], top-left origin), with an optional short label. Call this after "
    "cosmo_point_at_object rather than guessing a position yourself. Visual "
    "only — it measures nothing and changes nothing."
)

_DRAW_BOX_PARAMETERS: dict[str, Any] = {
    "type": "object",
    "properties": {
        "box": {
            "type": "object",
            "description": (
                "Where to draw, normalized to the frame you were shown: "
                "[0,1], top-left origin."
            ),
            "properties": {
                "x": {"type": "number", "minimum": 0, "maximum": 1},
                "y": {"type": "number", "minimum": 0, "maximum": 1},
                "width": {"type": "number", "minimum": 0, "maximum": 1},
                "height": {"type": "number", "minimum": 0, "maximum": 1},
            },
            "required": ["x", "y", "width", "height"],
        },
        "label": {
            "type": "string",
            "maxLength": 40,
            "description": "Short caption shown on the box, e.g. 'blush here'.",
        },
    },
    "required": ["box"],
}

_DRAW_POINT_PARAMETERS: dict[str, Any] = {
    "type": "object",
    "properties": {
        "point": {
            "type": "object",
            "description": (
                "Where to point, normalized to the frame you were shown: "
                "[0,1], top-left origin."
            ),
            "properties": {
                "x": {"type": "number", "minimum": 0, "maximum": 1},
                "y": {"type": "number", "minimum": 0, "maximum": 1},
            },
            "required": ["x", "y"],
        },
        "label": {
            "type": "string",
            "maxLength": 40,
            "description": "Short caption shown beside the marker, e.g. 'this screw'.",
        },
    },
    "required": ["point"],
}


@dataclass(frozen=True)
class NormalizedBox:
    """A rectangle in ``[0,1]``, **top-left origin** (y increases downward) —
    the convention the locators report and a preview overlay maps onto the
    screen."""

    x: float
    y: float
    width: float
    height: float


@dataclass(frozen=True)
class NormalizedPoint:
    """A position in ``[0,1]``, on the same terms as :class:`NormalizedBox`."""

    x: float
    y: float


@dataclass(frozen=True)
class DrawBoxRequest:
    """One model request to draw a box over the user's live view: where, and
    an optional short caption."""

    box: NormalizedBox
    label: str | None = None


@dataclass(frozen=True)
class DrawPointRequest:
    """One model request to mark a spot on the user's live view: where, and an
    optional short caption."""

    point: NormalizedPoint
    label: str | None = None


@dataclass(frozen=True)
class DrawOutcome:
    """What a renderer's handler reports back to the model.

    Drawing can fail for reasons the model must hear about — the camera is
    off, the preview isn't on screen, the frame the box describes is already
    gone. Answering ``shown=True`` regardless would leave the model talking
    about something the user cannot see, so a refusal travels with a
    ``reason``: model-facing prose the agent says out loud ("the camera is off
    — ask the user to turn it on"), not an error code.
    """

    shown: bool
    reason: str | None = None


DrawBoxHandler = Callable[
    [DrawBoxRequest], Union[DrawOutcome, Awaitable[DrawOutcome]]
]
"""Your box renderer: ``(request) -> outcome``, sync or async."""

DrawPointHandler = Callable[
    [DrawPointRequest], Union[DrawOutcome, Awaitable[DrawOutcome]]
]
"""Your point renderer: ``(request) -> outcome``, sync or async."""


def _clamped(raw: object) -> float | None:
    """A JSON number clamped into ``[0,1]``, or ``None`` when the model sent
    something that isn't one. Clamping keeps a model that overshoots the frame
    edge yielding a drawable annotation."""
    if isinstance(raw, bool) or not isinstance(raw, (int, float)):
        return None
    return min(1.0, max(0.0, float(raw)))


def _box(raw: object) -> NormalizedBox | None:
    if not isinstance(raw, dict):
        return None
    x, y = _clamped(raw.get("x")), _clamped(raw.get("y"))
    width, height = _clamped(raw.get("width")), _clamped(raw.get("height"))
    if x is None or y is None or width is None or height is None:
        return None
    return NormalizedBox(x=x, y=y, width=width, height=height)


def _point(raw: object) -> NormalizedPoint | None:
    if not isinstance(raw, dict):
        return None
    x, y = _clamped(raw.get("x")), _clamped(raw.get("y"))
    if x is None or y is None:
        return None
    return NormalizedPoint(x=x, y=y)


def _label(args: Mapping[str, Any]) -> str | None:
    label = args.get("label")
    return label if isinstance(label, str) else None


async def _tool_result(
    outcome: DrawOutcome | Awaitable[DrawOutcome],
) -> dict[str, Any]:
    resolved = outcome if isinstance(outcome, DrawOutcome) else await outcome
    result: dict[str, Any] = {"shown": resolved.shown}
    if resolved.reason is not None:
        result["reason"] = resolved.reason
    return result


def draw_box(on_draw: DrawBoxHandler) -> ClientTool:
    """The box renderer, ready to add to ``tools=`` alongside the locator that
    feeds it (:class:`~cosmo_ai.DetectObjectsTool`).

    ``on_draw`` receives a decoded, clamped :class:`DrawBoxRequest` and returns
    a :class:`DrawOutcome`; it may be sync or async. Malformed model arguments
    surface to the model as the invocation's error without reaching it.
    """

    async def handler(args: dict[str, Any]) -> dict[str, Any]:
        box = _box(args.get("box"))
        if box is None:
            raise ValueError(
                f"{DRAW_BOX_TOOL_NAME}: pass box {{x,y,width,height}} "
                f"normalized to [0,1]"
            )
        return await _tool_result(
            on_draw(DrawBoxRequest(box=box, label=_label(args)))
        )

    return _SdkClientTool(
        name=DRAW_BOX_TOOL_NAME,
        description=_DRAW_BOX_DESCRIPTION,
        parameters=_DRAW_BOX_PARAMETERS,
        handler=handler,
    )


def draw_point(on_draw: DrawPointHandler) -> ClientTool:
    """The point renderer, pairing with
    :class:`~cosmo_ai.PointAtObjectTool`. Same contract as :func:`draw_box`,
    with a :class:`DrawPointRequest` — it exists next to the box renderer
    because the two answer different questions: a box around a leaf includes
    everything behind it, where a marked point says one thing.
    """

    async def handler(args: dict[str, Any]) -> dict[str, Any]:
        point = _point(args.get("point"))
        if point is None:
            raise ValueError(
                f"{DRAW_POINT_TOOL_NAME}: pass point {{x,y}} normalized to [0,1]"
            )
        return await _tool_result(
            on_draw(DrawPointRequest(point=point, label=_label(args)))
        )

    return _SdkClientTool(
        name=DRAW_POINT_TOOL_NAME,
        description=_DRAW_POINT_DESCRIPTION,
        parameters=_DRAW_POINT_PARAMETERS,
        handler=handler,
    )
